source/MRViewer/MRSurfaceContoursWidget.cpp

```diff
@@ -0,0 +1,423 @@
+#include "MRSurfaceContoursWidget.h"
+#include "MRViewport.h"
+#include "MRMesh/MRHistoryAction.h"
+#include "MRAppendHistory.h"
+#include "MRMesh/MRMesh.h"
+#include "MRMesh/MRSphereObject.h"
+#include "MRMesh/MRObjectMesh.h"
+
+namespace MR
+{
+
+void updateBaseColor( std::shared_ptr<SurfacePointWidget> point, const Color& color )
+{
+    auto params = point->getParameters();
+    params.baseColor = color;
+    point->setParameters( params );
+}
+
+// History classes;
+
+std::string AddPointActionPickerPoint::name() const
+{
+    return "Add Point";
+}
+
+void AddPointActionPickerPoint::action( Type actionType )
+{
+    if ( !widget_.isPickerActive_ )
+        return;
+
+    auto& contour = widget_.pickedPoints_[obj_];
+    if ( actionType == Type::Undo )
+    {
+        contour.pop_back();
+
+        if ( !contour.empty() )
+            updateBaseColor( contour.back(), widget_.params.lastPoitColor );
+
+        widget_.activeIndex = int( contour.size() );
+        widget_.activeObject = obj_;
+
+        widget_.onPointRemove_( obj_ );
+    }
+    else
+    {
+        if ( !contour.empty() )
+            updateBaseColor( contour.back(), widget_.params.ordinaryPointColor );
+
+        contour.push_back( widget_.createPickWidget_( obj_, point_ ) );
+
+        updateBaseColor( contour.back(), widget_.params.lastPoitColor );
+
+
+        widget_.onPointAdd_( obj_ );
+    }
+
+}
+size_t AddPointActionPickerPoint::heapBytes() const
+{
+    return 0; //this undo action will be deleted in widget disable
+}
+
+
+
+std::string RemovePointActionPickerPoint::name() const
+{
+    return "Remove Point";
```

Comment:
please add widget name here, not just remove points, or add option for user to specify these names

Reply:
add history specification params